### PR TITLE
CDAP-5180 add conversions between config versions

### PR DESCRIPTION
**Pull request by albertshau**
Moves the configs from 3.2.x from the upgrade tool into the proto
module. Also moves the upgrade logic from them and formalizes the
upgrade process so that each config knows how to upgrade to the
next version, and upgrades can be chained together to reach the
final version.

Adding upgrade logic from v1 of the configs to v2, which will
allow us to internally use v2 everywhere when we implement the
upcoming workflow planner.

### CODE REVIEW

**Comment by albertshau** [comment]
http://builds.cask.co/browse/CDAP-DUT3641

**Comment by albertshau** [comment]
also looks like a lot, but most of it is moving OldETL* classes and tests from the upgrade tool into proto as v0 configs.  The rest is adding logic to upgrade v1 configs into v2 configs.

File: cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/co/cask/cdap/etl/proto/UpgradeableConfig.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+/**
+ * Upgradeable config for ETL applications, which allows chain upgrading until the latest version is reached.
+ *
+ * @param <T> type of config it gets upgraded to.
+ */
+public interface UpgradeableConfig<T extends UpgradeableConfig> {
+
+  boolean canUpgrade();
+
+  T upgrade(UpgradeContext upgradeContext);
```
Comment (gsps1) [inline cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/co/cask/cdap/etl/proto/UpgradeableConfig.java]: need description for these methods.

File: cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/co/cask/cdap/etl/proto/UpgradeableConfig.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.etl.proto;
+
+/**
+ * Upgradeable config for ETL applications, which allows chain upgrading until the latest version is reached.
```
Comment (gsps1) [inline cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/co/cask/cdap/etl/proto/UpgradeableConfig.java]: more explanation with example would help.

File: cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/co/cask/cdap/etl/proto/v0/ETLRealtimeConfig.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+import co.cask.cdap.etl.api.Transform;
+import co.cask.cdap.etl.api.realtime.RealtimeSink;
+import co.cask.cdap.etl.api.realtime.RealtimeSource;
+import co.cask.cdap.etl.proto.UpgradeContext;
+import co.cask.cdap.etl.proto.UpgradeableConfig;
+
+import java.util.List;
+
+/**
+ * ETL Realtime Configuration.
```
Comment (gsps1) [inline cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/co/cask/cdap/etl/proto/v0/ETLRealtimeConfig.java]: sample configuration here would be useful as we look back..

Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/co/cask/cdap/etl/proto/v0/ETLRealtimeConfig.java]: this is just normal serialization/deserialization like we use everywhere in the code base. I would rather not put an example in every single class that is just a json of its fields...

Comment (gsps1) [inline cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/co/cask/cdap/etl/proto/v0/ETLRealtimeConfig.java]: yea ok

File: cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/co/cask/cdap/etl/proto/v0/ETLRealtimeConfig.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+      builder.addConnection(prevStageName, currStageName);
+      prevStageName = currStageName;
+      stageNum++;
+    }
+
+    for (ETLStage v0Sink : getSinks()) {
+      String currStageName = v0Sink.getName() + "." + stageNum;
+      builder.addSink(v0Sink.upgradeStage(currStageName, RealtimeSink.PLUGIN_TYPE, upgradeContext));
+      builder.addConnection(prevStageName, currStageName);
+      stageNum++;
```
Comment (gsps1) [inline cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/co/cask/cdap/etl/proto/v0/ETLRealtimeConfig.java]: most of this code is common with ETLBatchConfig, can you push this to ETLConfig as both of them extend it and remove duplication ?

File: cdap-app-templates/cdap-etl/cdap-etl-tools/src/main/java/co/cask/cdap/etl/tool/UpgradeTool.java [hunk 5]
```diff
@@ -481,7 +324,44 @@ private static boolean shouldUpgrade(ArtifactSummary artifactSummary) {
     // check its 3.2.x versions of etl
     ArtifactVersion artifactVersion = new ArtifactVersion(artifactSummary.getVersion());
     return !(artifactVersion.getMajor() == null || artifactVersion.getMinor() == null) &&
-      artifactVersion.getMajor() == 3 && artifactVersion.getMinor() == 2;
+      artifactVersion.getMajor() == 3 && (artifactVersion.getMinor() == 2 || artifactVersion.getMinor() == 3);
```
Comment (gsps1) [inline cdap-app-templates/cdap-etl/cdap-etl-tools/src/main/java/co/cask/cdap/etl/tool/UpgradeTool.java]: should this be `(artifactVersion.getMinor() >= 2)` if we want to support other minor releases

**Comment by gsps1** [comment]
just few comments, LGTM otherwise, do we have an upgrade integration test covering this?

File: cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/co/cask/cdap/etl/proto/v0/ETLRealtimeConfig.java (hunk 4 above)
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/co/cask/cdap/etl/proto/v0/ETLRealtimeConfig.java]: good point, will do

File: cdap-app-templates/cdap-etl/cdap-etl-tools/src/main/java/co/cask/cdap/etl/tool/UpgradeTool.java (hunk 5 above)
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-tools/src/main/java/co/cask/cdap/etl/tool/UpgradeTool.java]: also a good point, will change.

**Comment by albertshau** [comment]
there is no integration test right now, though we should add one. will file a jira

**Pull request merged**